pkg/operators/ebpf/stats.go

```diff
@@ -0,0 +1,535 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ebpfoperator
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/cilium/ebpf"
+	"github.com/spf13/viper"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/bpfstats"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	metadatav1 "github.com/inspektor-gadget/inspektor-gadget/pkg/metadata/v1"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+const (
+	ParamGadgetStatisticsInterval = "statistics-interval"
+	ParamGadgetStatisticsCount    = "statistics-count"
+	EmitStatsAnn                  = "emitstats"
+	ParamProgramStats             = "programs"
+	StatsDSName                   = "bpfstats"
+	EnableBPFStatsParam           = "enable-bpfstats"
+)
+
+func (i *ebpfOperator) GlobalParams() api.Params {
+	return api.Params{
+		{
+			Key:          EnableBPFStatsParam,
+			Description:  "Enable capturing eBPF stats",
+			DefaultValue: "false",
+			TypeHint:     api.TypeBool,
+			Title:        "Enable eBPF stats",
+		},
+	}
+}
+
+func (i *ebpfOperator) Init(params *params.Params) error {
+	if !params.Get(EnableBPFStatsParam).AsBool() {
+		return nil
+	}
+
+	i.gadgetObjs = make(map[operators.GadgetContext]gadgetObjs)
+
+	// Enable stats collection
+	// TODO: when to disable it?
+	if err := bpfstats.EnableBPFStats(); err != nil {
+		return fmt.Errorf("enabling bpf stats: %w", err)
+	}
+
+	return nil
+}
+
+func (i *ebpfOperator) InstantiateDataOperator(
+	gadgetCtx operators.GadgetContext, paramValues api.ParamValues,
+) (operators.DataOperatorInstance, error) {
+	cfg, ok := gadgetCtx.GetVar("config")
+	if !ok {
+		return nil, fmt.Errorf("missing configuration")
+	}
+	v, ok := cfg.(*viper.Viper)
+	if !ok {
+		return nil, fmt.Errorf("invalid configuration format")
+	}
+
+	emitStatsAnn := v.GetBool(fmt.Sprintf("annotations.operator.%s.%s", i.Name(), EmitStatsAnn))
+	if !emitStatsAnn {
+		return nil, nil
+	}
+
+	if i.gadgetObjs == nil {
+		return nil, fmt.Errorf("bpfstats aren't enabled")
+	}
+
+	var err error
+
+	instance := &ebpfOperatorDataInstance{
+		bpfOperator:  i,
+		done:         make(chan struct{}),
+		programStats: paramValues[ParamProgramStats] == "true",
+	}
+
+	instance.ds, err = gadgetCtx.RegisterDataSource(datasource.TypeArray, StatsDSName)
+	if err != nil {
+		return nil, err
+	}
+
+	intervalAnn := paramValues[ParamGadgetStatisticsInterval]
+	if intervalAnn == "" {
+		intervalAnn = "1000ms"
```

Comment:
This should be already there, isn't it?

Reply:
Mmm you can simply don't validate them if they are not present. That's what I did in the [limiter](https://github.com/inspektor-gadget/inspektor-gadget/blob/4d0d32998539152e1156ce3a2cf5fafa15c61cea/pkg/operators/limiter/limiter.go#L67-L80).

Reply:
I found an issue with my approach in the limiter. There are some cases, like for unit tests, where the operator is called without `paramValues`. If I don't set a default value, then the operator will run with `0`, which in the case of the limiter is to truncate all the output 😅.

Reply:
Updated following #4050